은행창구 관리앱 [STEP 3] Blue, Lust3r

llimental opened the pull request:
안녕하세요! Corn(@protocorn93)🍿!
💫**TEAM BLUST3R**💫의 Blue, Lust3r입니다!

날씨가 많이 따뜻해지긴 했지만 일교차가 엄청 커졌어요!
감기 조심하세요 콘! 🤧🤧

이번 Step의 요구사항인 다중처리를 구현하기 위해
동시성 프로그래밍, 동기/비동기에 대한 학습을 선행했습니다.
어려운 개념인 만큼 아직 많이 부족하지만 최선을 다해 Step 진행해보았습니다.

이번 Step 리뷰도 잘 부탁드립니다! 감사합니다! 🙇🏻🙇🏻

## 🧑🏼‍💻 작업 목록

**[ STEP 3 ] 다중처리**

- [x] 3명의 은행원 근무
    - [x] 2명 예금업무 / 1명 대출업무 처리
    - [x] 고객 업무 종류는 고객 생성시 임의 지정
    - [x] 한 명의 은행원은 한 명의 고객을 응대
- [x] 업무 종류 및 소요시간 추가 구현
    -  대출: 1.1초
    -  예금: 0.7초
- [x] 업무처리를 시작 시 메세지 출력
    - "3번 고객 대출업무 시작"
- [x] 업무 처리 완료 시 메세지 출력
    - "11번 고객 예금업무 완료"


## 📚 학습 키워드 및 학습 내용

### 1. DispatchQueue: Sync/Async

>[출처: Apple 공식문서 - DispatchQueue](https://developer.apple.com/documentation/dispatch/dispatchqueue)

- **Sync, 동기**
    - 앞선 작업이 종료될 때까지 대기 후 다음 작업 처리
- **Async, 비동기**
    - 앞선 작업이 종료되기를 기다리지 않고 다른 스레드에서 처리
- **Sync, Async의 차이**
    - 실행 종료 시점을 알 수 있는가에 대한 차이
    - 비동기의 경우 언제 작업이 종료되는지 알 수 있음
- **DispatchQueue**
    - 동기 프로그래밍을 하면 이전 코드가 완료될 때 까지 기다려야 해서 프로그램의 지연을 발생시킨다는 단점이 있음
이 때문에 비동기 프로그래밍을 하면 오래 걸리는 코드를 별도로 처리하면서도 다른 작업을 하는 효율을 꾀할 수 있는 장점이 있음
    - `main().async`
        - main스레드에서 비동기 작업
        - 다른 스레드를 만들지 않고 main 스레드에서만 일을 처리
        - 작업이 끝나기를 기다리지는 않고 있지만 main 스레드라는 단일 스레드에서만 작업이 이루어지고 있기 떄문에 동시에 작업이 처리되지는 못함
        - 쌓인 순서대로 작업 처리
    - `global().async`
        - main 스레드가 아닌 다른 스레드를 만들어 비동기적으로 작업 처리
        - async의 특성상 어떤 코드가 먼저 실행될 지는 예측할 수 없음
        - 각 스레드마다 작업이 처리되는 속도가 다를 수 있고, 이를 직접 통제할 수 없음
        - 동시에 처리되는 작업에 대해서는 논리적으로 정확한 순서를 기대하기는 어려움
    - `global().sync`
        - main 스레드가 아닌 다른 스레드를 생성하고 그 위에서 동기적으로 작업을 처리해주느 메서드
        - 스레드는 각각 다르지만 동기적으로 일을 처리하기 때문에 각각의 작업이 끝나기를 기다림
    - `main().sync`
        - main 스레드에서 직접 호출하면 안되는 코드
        - 작업이 끝나기를 기다리는 sync의 특성 떄문에 직접적으로 호출하는 경우 교착상태(deadlock)에 빠지게 됨
        - Serial 큐를 커스텀하여 sync를 실행하면 에러는 발생하지 않음
        - global 스레드에서 호출하는 것은 괜찮다!
- **DispatchGroup**
    - 작업 그룹을 만들어 그 그룹을 하나의 단위로 모니터링할 때 사용
    - group.wait() 메서드를 DispatchGroup의 수행이 끝나기를 기다릴 수 있음

🌟 이번 스텝에서는 **은행원 3명**이 **비동기적으로 업무를 처리**할 수 있도록 `DispatchQueue.global().async`를 사용하여 구현하였음
또한, 각 업무(예금/대출)에 해당하는 비동기 작업들을 `DispatchGroup`으로 묶어 해당 그룹의 업무가 모두 처리되기를 `wait()` 메서드를 사용하여 기다렸다가 업무 종료 메세지를 출력하도록 함


### 2. Escaping closure
> A closure is said to escape a function when the closure is passed as an argument to the function, but is called after the function returns.
[출처: swift 공식문서 - Closures](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/closures/#Escaping-Closures)
- closure는 해당 코드 구문에서 진행되어야 하지만 Escaping closure는 그 코드 구문, 함수가 반환된 이후에 진행되는 차이점이 있음.
- 또한 closure는 정의된 주변 컨텍스트에서 값을 캡처하는데 상수와 변수에 대한 참조를 캡처하고 저장할 수 있어 원래 범위가 더 이상 존재하지 않아도 본문 내에서 해당 상수와 변수의 값을 참조하고 수정할 수 있음.
- 하지만 escaping closure는 캡쳐한 값이 복사된 것이기에 그 안에서 수정하게 되더라도 원본에 영향을 줄 수 없어 mutable한 struct를 사용할 수 없고 class를 써야 함.
공식문서에서도 struct와 enum은 shared mutability를 허용하지 않기에 mutable한 참조는 캡처할 수 없다고 함.
자세한 내용은 문서에 나와있지 않아 설명을 바탕으로 추측해보면, closure는 해당 구문 내에서 실행되고, mutating을 통해 원본 값에 반영을 할 수 있지만 escaping closure는 해당 구문을 벗어나기 때문에 값을 캡처해서 복사하기 때문이라고 생각함.
이 캡처도 reference type, value type을 구분하기 때문에 value type이라면 'Structures and Enumerations Are Value Types'에서 말하듯 값이 변경된 순간 다른 복사본 메모리로 만들어지기 때문에 원본에 영향이 없지만, reference type이라면 복사된 값이 변경되더라도 원본에 영향을 줄 수 있기 때문에 차이가 있다고 생각함.


### 3. Semaphore
> An object that controls access to a resource across multiple execution contexts through use of a traditional counting semaphore.
[출처: apple 공식문서 - Dispatch Semaphore](https://developer.apple.com/documentation/dispatch/dispatch_semaphore/)
- 공유자원에 접근할 수 있는 스레드의 수를 제어해 줌으로써 Race Condition을 방지해주는 역할도 함
- 허용된 스레드 수만큼 접근된 상태라면 다른 스레드는 접근하지 못하고 줄을 서서 기다리게 됨

🌟 이번 스텝에서는 `CustomerQueue`에서 Customer를 `dequeue`한 후, 그 값에 접근하는 thread의 수를 제어해서 예금 업무에는 2개의 스레드가 접근할 수 있도록, 대출 업무인 경우 1개의 스레드만 접근할 수 있도록 구현하여 Step3의 요구사항인 예금 은행원 2명, 대출 은행원 1명을 적용하였음

## 🤔 고민과 해결
### 1. 3명의 은행원 비동기 처리
- Step 3을 처음 설계 시 은행원이 3명이라는 구현 요구사항을 보고 은행원 객체를 3개 생성하여야 할까? 라는 고민을 하였습니다. 하지만 저번 step에서 콘이 조언해주신 것처럼 `프로그램의 서비스 시나리오와 프로그램의 시나리오가 동일시 되는 것 같다`라는 생각에 러스터와 함께 고민해 보았습니다.
- 학습 키워드 중 `동시성 프로그래밍 개념의 이해`, `thread의 개념에 대한 이해`가 있어 설계 전 해당 부분을 충분히 학습하기로 결정하고 학습 결과, 은행원을 인원수에 맞춰 객체로 생성하는 것보다 업무를 구분하고, 해당 업무에 접근할 수 있는 thread의 수로 은행원의 수를 결정하는 것이 좋겠다 생각하였고, 이를 구현하기 위해 업무별로 semaphore를 선언하여 접근할 수 있는 스레드 수를 제어해 주었습니다.

    ```swift
    // Bank.swift 의 open() 메서드 내부
        // ...
        let loanSemaphore = DispatchSemaphore(value: 1)
        let accountSemaphore = DispatchSemaphore(value: 2)

        while !customerQueue.isEmpty() {
            // ...
            switch currentCustomerWorkType {
            case WorkList.account.rawValue :
                DispatchQueue.global().async(group: group) {
                    accountSemaphore.wait()
                    accountBanker.work(of: currentCustomer.waitingOrder, for: currentCustomerWorkType)
                    todayCounter.addAccountCustomer()
                    accountSemaphore.signal()
            }
            default:
                DispatchQueue.global().async(group: group) {
                    loanSemaphore.wait()
                    accountBanker.work(of: currentCustomer.waitingOrder, for: currentCustomerWorkType)
                    todayCounter.addLoanCustomer()
                    loanSemaphore.signal()
            }
        }
    }
    ```

### 2. 메서드 역할 분리(Counter)
- 리팩토링 전 작성했던 Bank의 `open()`메서드 내부에 프로퍼티가 많이 포함되어 있고, 한 객체에 너무 많은 역할이 주어지는 것 같아서 역할 분리를 진행해 보고자 하였습니다.
    ```swift
    // refactoring 이전

    let numberOfTodayCustomers = makeAcceptableNumber()
    let customerQueue = makeCustomerQueue(maxCount: numberOfTodayCustomers)
    let loanSemaphore = DispatchSemaphore(value: 1)
    let accountSemaphore = DispatchSemaphore(value: 2)
    let group = DispatchGroup()
    var numberOfAccountCustomers = 0
    var numberOfLoanCustomers = 0
    ```
- 위의 프로퍼티를 보고 분리할 수 있는 역할에 대해 고민해보다가 고객의 수를 집계하는 역할을 분리해보고자 Counter 클래스를 구현하였습니다.
- Counter 클래스는 예금 고객의 수 / 대출 고객의 수를 프로퍼티로 가지고 각 고객을 1씩 증가시키는 add, 각 고객의 수를 가져오는 get, 모든 고객의 수를 초기화 시키는 reset 메서드를 가지고 있습니다.
- 역할 분리로 인해 Bank 객체가 한 가지 역할을 덜 수행할 수 있게 되었고, `open()` 메서드 내부에서 처리해야 하는 일도 다른 객체에 위임할 수 있게 되었습니다!


## 🙋🏻 질문 거리
### 1. Escaping Closure의 Capture 원리
- Customer의 수를 세기 위해 Counter를 class 타입으로 만들었습니다. 그런데 struct와 class 중에 선택할 때 class를 선택해야 하는 이유가 없다면 대부분의 경우에는 struct를 권장한다는 문서의 말에 따라 struct로 한 번 변경해보니 escaping closure에서 mutable한 값을 캡처할 수 없다는 오류를 보게 되었습니다.
- Closure 공식문서를 통해 escaping closure는 mutable한 struct, enum은 capture할 수 없다는 설명은 보았지만, 그 이유에 대해서는 찾을 수 없었습니다.
- 다방면으로 찾아보고 주변에 물어본 결과, 값을 복사해서 사용하기 때문에 escaping closure가 끝나면 원본에 반영되지 않고 사라지기에 그렇다-라는 결론이 나왔습니다. 그렇다면 값타입과 참조타입의 차이가 반영이 되었다는 것인데, 클로저 공통 설명에서는 값을 '참조'한다고 되어 있어 값타입이든 참조타입이든 변경이 가능한 것이 아닐까 하는 생각도 들었습니다.
- 위의 결론과 생각을 다 만족시키려면 closure와 escaping closure의 캡처 방식이 다르거나 문서에서의 reference가 type을 뜻하는 것이 아니라 그냥 값 / 주소를 참고 한다의 '참조'이거나 둘 중 하나일 것이라 생각했습니다.
- escaping closure에서 mutable한 값은 캡처할 수 없다는 문구와, 값을 캡처(복사)하여 사용해서 그렇다는 설명을 보고 납득은 했지만 이러한 이해가 맞는 것인지 여쭤보고 싶습니다!

### 2. `if` vs `guard`
- `Bank.swift`의 `open()` 메서드 내부에서 고객을 dequeue한 후 고객의 업무에 따라 분기처리를 할 수 있게끔 구현하려고 하는 부분에서 원래 `if-else`를 사용하였습니다.

    ```swift
    if currentCustomerWorkType == WorkList.account.rawValue {
        DispatchQueue.global().async(group: group) {
            accountSemaphore.wait()
            accountBanker.work(for: currentCustomer.waitingOrder, what: currentCustomerWorkType)
            todayCounter.addAccountCustomer(with: 1)
            accountSemaphore.signal()
        }
    } else {
        DispatchQueue.global().async(group: group) {
            loanSemaphore.wait()
            loanBanker.work(for: currentCustomer.waitingOrder, what: currentCustomerWorkType)
            todayCounter.addLoanCustomer(with: 1)
            loanSemaphore.signal()
        }
    }
    ```
- 클린코드 스터디를 진행한 버드에게서 `if-else` 구문을 사용하게 되면 `guard` 구문을 사용했을 때보다 가독성이 떨어진다는 이야기를 듣게 된 후 `guard-else`로 변경해 보았으나 guard는 조건에 일치하지 않는 경우 특정 실행 구문을 빠르게 종료하는 Early-exit의 특성이 강하다고 생각되었습니다.

- 추가적으로 학습해본 결과 if문에서 else 사용을 지양하자! 라고 주장하는 이유는 아래의 두 가지로 추정되었습니다
    - style(가독성) 관련
    - 반전된 로직을 작성하게 되는 논리적 위험 발생 가능성
- 저희의 프로젝트, Step 3에서는 업무가 예금 / 대출의 두 가지 중 하나이기 때문에 논리적 위험 발생 가능성은 없지만, else로 인해 해당 부분을 들여쓰게 되어 가독성 관련하여서는 지양해야 하는 케이스가 될 것 같습니다.
- 저희의 경우 if-else를 사용해도 무방해 보이지만, guard-else{continue}를 써도 괜찮을 거 같아요. 현업에서 가독성 때문에 if문 사용 시 else 사용을 지양하는 편인지, 콘은 어떻게 사용하는 게 더 좋을 것 같은지가 궁금합니다!


### 3. DispatchQueue 관련 질문
- semaphore를 account와 loan으로 두 개 사용하고, group까지 설정하니 선언한 프로퍼티가 벌써 3개가 넘어가 기본 타입 3개 이하로 유지해야 하는 원칙을 위배하게 되었습니다.
- 비동기 프로그래밍을 하게 된다면 이처럼 설정해줘야 하는 것이 많은데 semaphore나 이런 프로퍼티도 따로 빼내서 사용할 수 있는지 여부와 저희 은행원 쓰레드의 async 코드도 사용하는 semaphore만 빼면 비슷한 구성인데 하나로 묶어서 사용할 수 있는 방법이 있을지 궁금합니다!
    ```Swift
    let loanSemaphore = DispatchSemaphore(value: 1)
    let accountSemaphore = DispatchSemaphore(value: 2)
    let group = DispatchGroup()

    while !customerQueue.isEmpty() {
        guard let currentCustomer = customerQueue.dequeue() else { return }

        let currentCustomerWorkType = currentCustomer.workType

        switch currentCustomerWorkType {
        case WorkList.account.rawValue :
            DispatchQueue.global().async(group: group) {
                accountSemaphore.wait()
                accountBanker.work(of: currentCustomer.waitingOrder, for: currentCustomerWorkType)
                todayCounter.addAccountCustomer()
                accountSemaphore.signal()
            }
        default:
            DispatchQueue.global().async(group: group) {
                loanSemaphore.wait()
                accountBanker.work(of: currentCustomer.waitingOrder, for: currentCustomerWorkType)
                todayCounter.addLoanCustomer()
                loanSemaphore.signal()
            }
        }
    }
    ```

calledBlu committed:
[doc] Update README

calledBlu committed:
[doc] Update README

add newline

llimental committed:
[feat] WorkList 정의 및 Customer에 프로퍼티 추가

- WorkList: 예금, 대출 정의
- Customer: workType 프로퍼티 추가
- CustomerManageable: setRandomWork로 workType을 받아 enqueue 작업

calledBlu committed:
[feat] Banker 타입 확장 및 수정

- 은행원이 예금, 대출 업무를 수행할 수 있도록 수정
- Banker가 정해진 processingTime을 가지지 않고 초기화 시 processingTime을 정할 수 있도록 수정
- Banker의 메서드가 처리하는 workType을 파라미터로 받을 수 있도록 프로토콜 수정
- 프로토콜 수정에 따른 Banker, Bank 일부 코드 수정

llimental committed:
[feat] DispatchQueue를 사용하여 비동기로 Customer 업무 처리

- Semaphore를 사용하여 각 업무별 은행원 스레드 제한
- group을 사용하여 모든 은행원이 업무가 끝난 이후에 메세지가 출력되도록 설정
- calculateTotalWorkTime() 메서드를 통해 각 업무별 customer 수와 시간을 합한 값 반환

calledBlu committed:
[feat] 업무 처리 시 고객의 수를 관리하는 Counter 구현

- Bank의 numberOfAccountCustomers, numberOfLoanCustomers를 Counter로 이관
- 예금/대출 고객의 수를 증가시키고, 가져오고, 초기화하는 메서드 구현
- Bank에 상기 내용을 반영하여 코드 수정

llimental committed:
[refactor] 코드 리팩터링

- Bank: if else 구문을 switch case로 변경
- Counter: 상속하지 않는 클래스여서 final 키워드 사용
- Counter: addCustomer 메서드 코드를 변경함으로써 메서드 호출시 매직넘버를 사용하지 않도록 수정

calledBlu committed:
[fix] 코드 오류 정정

- accountBanker -> loanBanker로 수정

protocorn93 commented on BankManagerConsoleApp/BankManagerConsoleApp/Enum/WorkList.swift:
```diff
@@ -0,0 +1,21 @@
+
+import Foundation
+
+enum WorkList: String, CaseIterable {
```
`-List` 네이밍보단 `WorkType`이 더 알맞은 네이밍 같아보여요

llimental replied:
의견 감사해요 콘! 네이밍 반영하였습니다!

해당 커밋: 3914fbd74b6870c9c2163f432505c63d438b27ab

protocorn93 commented on BankManagerConsoleApp/BankManagerConsoleApp/Enum/WorkList.swift:
```diff
@@ -0,0 +1,21 @@
+    case account = "예금"
+    case loan = "대출"
+
+    static func setRandomWork() -> String {
```
`set-` 네이밍보단 함수의 역할이 getter이기 때문에 `get-` 네이밍을 사용하는게 좋을 것 같아요. 

그리고 이와 별개로 랜덤한 케이스를 반환해주는 것은 이 타입의 역할은 아니라고 생각해요. 사용하는 곳에서 generate하는게 좋을 것 같아요.

calledBlu replied:
해당 메서드 이름을 `getRandomWork()`로 변경하고, 각 객체의 역할을 생각해본 후 해당 메서드의 위치를 `WorkType` -> `CustomerManageable`으로 변경하였어요!

> 해당 커밋: 7026e6eb9e3d1087946b803b8cfa19c97fb45ca1

protocorn93 commented on BankManagerConsoleApp/BankManagerConsoleApp/Model/Bank.swift:
```diff
@@ -8,28 +8,63 @@
+            case WorkList.account.rawValue :
+                DispatchQueue.global().async(group: group) {
+                    accountSemaphore.wait()
+                    accountBanker.work(of: currentCustomer.waitingOrder, for: currentCustomerWorkType)
```
그냥 `Customer` 객체를 바로 받아도 되지 않을까요

llimental replied:
work 메서드 로직을 수정해서, currentCustomer의 값을 파라미터로 던져주는 것이 아니라, Customer를 보내서 work에서 그 값을 사용할 수 있도록 변경했습니다!
덕분에 코드도 좀 더 보기 편하게 변경되었습니다. 감사해요 콘!

해당 커밋: e3289246d83f1fc9866249ef636119c60aa69dc2

protocorn93 commented on BankManagerConsoleApp/BankManagerConsoleApp/Model/Bank.swift:
```diff
@@ -8,28 +8,63 @@
-    let banker = Banker()
-    let processingTimePerPerson = Banker.processingTime
+
+    let todayCounter = Counter()
+    let accountBanker = Banker(processingTime: 0.7)
+    let loanBanker = Banker(processingTime: 1.1)
```
조금 욕심 같아선 생성자로 주입 받고 싶네요!

calledBlu replied:
`Bank`의 생성자를 통해 주입될 수 있도록 변경하고, `processingTime`을 할당할 때 매직넘버를 사용하지 않도록 리팩토링 진행해 보았습니다.

> 해당 커밋: 6f9e4d17e95a47468de6197f331bce584cb84450

protocorn93 commented on BankManagerConsoleApp/BankManagerConsoleApp/Model/Bank.swift:
```diff
@@ -8,28 +8,63 @@
         print("업무가 마감되었습니다. 오늘 업무를 처리한 고객은 총 \(totalNumber)명이며, 총 업무시간은 \(convertDoubleToString)초입니다.")
     }
+
+    func calculateTotalWorkTime(customers: Int..., processingTimes: Double...) -> Double {
+        var result = 0.0
+
+        customers.enumerated().forEach {
+            let partialResult = Double($1) * processingTimes[$0]
+            result += partialResult
+        }
+
+        return result
+    }
```
Timer 객체를 사용해서 업무 개시할 때 시작하고 DispatchGroup의 모든 일이 끝나면 stop 해서 걸린 시간을 계산해보는건 어떨까요?

llimental replied:
Timer를 사용해보고자 했으나 학습해야 할 양이 많아 지금 상태로는 코드를 작성하는 것이 어렵다고 판단했어요..🥲
그래서 비동기 처리 전 현재 시간을 측정하고, 처리가 끝난 후 시간과 처음 시간의 차이를 구해 전체 작업에 걸린 시간을 출력하는 방식으로 진행해봤어요!
나중에 충분한 학습 후 가능하다면 블루와 Timer로 리팩터링 해보도록 할게요!

해당 커밋: ee20ed76562ba4fb1c48d0ccb63d68103340932a

protocorn93 commented on BankManagerConsoleApp/BankManagerConsoleApp/Model/Counter.swift:
```diff
@@ -0,0 +1,34 @@
+
+import Foundation
+
+final class Counter {
+    private var numberOfAccountCustomers = 0
+    private var numberOfLoanCustomers = 0
+
+    func addAccountCustomer() {
+        numberOfAccountCustomers += 1
+    }
+
+    func addLoanCustomer() {
+        numberOfLoanCustomers += 1
+    }
+
+    func getAccountCustomerCount() -> Int {
+        return numberOfAccountCustomers
+    }
+
+    func getLoanCustomerCount() -> Int {
+        return numberOfLoanCustomers
+    }
+
+    func resetCounter() {
+        numberOfAccountCustomers = 0
+        numberOfLoanCustomers = 0
+    }
+}
```
개인적으론 이 Counter 객체의 존재 의미를 모르겠어요. `makeAcceptableNumber` 반환된 결과를 그대로 사용하면 되지 않나요?

llimental replied:
의견 반영하여 리팩토링 해보았습니다!
Customer 수를 사용하는게 시간 계산을 위해서였는데, timeInterval로 측정하게 되면서 해당 코드들도 없앨 수 있었습니다.
감사해요 콘!

해당 커밋: c626bbcfa446781069a56972ed382dac7e8a2972

protocorn93 commented on BankManagerConsoleApp/BankManagerConsoleApp/Enum/WorkList.swift:
```diff
@@ -0,0 +1,21 @@
+
+enum WorkList: String, CaseIterable {
+
+    case account = "예금"
+    case loan = "대출"
```
`rawValue`를 활용하기보단 `description`이라는 변수를 사용하는건 어떨까요? 

보통 현업에서 `enum`의 `rawValue`는 네트워크 응답값에 매핑할 때 사용하곤해요.

calledBlu replied:
description을 사용할 수 있도록 CustomStringConvertible을 채택하였고,
rawValue를 쓰지 않게 되어 String 프로토콜 채택을 취소하였습니다!

> 해당커밋: 1a96868e4801b52a3fb7c71c1ceaf74b35503fc2

protocorn93 reviewed (changes requested):
동시성이라는게 쉽지 않은 스탭이었는데, 두 분 고생 많으셨어요! 학습을 많이 하셨다는게 느껴지는 PR이었어요. 

## 질문거리 

>  Escaping Closure의 Capture 원리

closure와 escaping closure의 차이는 실행시점입니다. 일반적인 closure는 메서드 안에서 바로 실행됩니다. 하지만 메서드 안에서 바로 실행되지 않고, 변수에 저장되거나, 언제 실행될지 모르는 closure라면 현재 문맥을 벗어나서 실행된다는 의미에서 escaping closure라 불립니다. 

struct, enum과 같은 값타입의 인스턴스는 설계자체가 immutable 합니다. 애초에 기대하는 바가 복사이지 변경이 아니기 때문이죠. 복사는 말그대로 복사이기 때문에 이전 값의 영향을 주지 않아요. 이게 우리가 값타입에 기대하는바에요. 이를 통해 프로그램 흐름에 있어 예상치 못한 값의 변경을 방지하는데 사용할 수 있죠. 

escaping closure에서 값을 캡처한다는 것은 언젠가 이 closure가 실행될 때 캡처한 값을 읽거나 쓴다는 것을 의미하죠. 역시 쓴다는 의미가 내포되어 있기 때문에 값타입을 캡처한다는 것은 앞뒤 논리가 성립이 되지 않는것이죠.

> if vs guard

저는 이거 혹은 저거를 수행한다는 의미에서 `if-else`를 사용해요. "이럴 경우만" 수행해! 라면 `guard`를 사용하죠. 늬앙스가 조금 다른 것을 느낄 수 있을거에요.

early exit은 if문으로도 충분히 가능하죠. 참 이게 어떤게 정답이다라곤 할 수 없는 것 같아요. 동료들끼리 어떤 합의가 되어있으며 어떤 방식으로 코드를 작성하는지, 어떤 방식을 사용해야 동료들이 내 코드를 읽을 때 좀 더 직관적일 수 있을까에 대해서 고민해보면 좋을거에요.


>  DispatchQueue 관련 질문

전 개인적으론 BankerProtocol이 존재해서 아래와 같은 그림을 상상했어요. 

```swift 
final class LoanBanker: BankerProtocol { 

}

final class AccountBanker: BankerProtocol { 
  private let semaphore: DispatchSemaphore 
  ... 

  init(numberOfBankers: Int, ... ) { 
    self.semaphore = DispatchSemaphore(value: numberOfBankers)
    ...
  }

  func work(customer: Customer) { 
    guard customer.workType == .account else { return } 
    ... 
  } 
}

let bankers: [BankerProtocol]

init(bankers: [BankerProtocol]) {
  self.bankers = bankers
}

customers.forEach { customer in 
  bankers.forEach { $0.work(customer) }
}

```

대충 적은 코드이지만 대략 저런 모습을 상상했어요. Bank에서 DispatchGroup을 각각 주입해주고 모든 일이 끝나면 notify 해주고 이때 Timer가 경과한 시간을 측정하는 그런 모습을 상상했습니다.

llimental committed:
[chore] WorkList -> WorkType 네이밍 변경

calledBlu committed:
[refactor] 메서드 네이밍 변경 및 이동

- 메서드의 역할에 맞게 set -> get으로 네이밍 변경
- 랜덤한 work type를 반환하는 매서드의 위치 변경 WorkType -> CustomerManageable
- 이전 commit의 네이밍 변경으로 인한 호출부분 코드 수정

llimental committed:
[refactor] 사용하지 않는 Counter class 제거

- Counter 사용 메서드가 시간 출력 관련된 메서드인데, 사람 수를 카운팅하지 않아도 구현 가능
- Counter 관련 메서드 제거(addAccountCustomer, resetCounter)
- calculateTotalWorkTime 메서드의 파라미터 수정(추후 Timer로 대체)

calledBlu committed:
[refactor] rawValue를 사용하지 않도록 수정

- WorkType에서 case에 해당하는 String 값을 가져올 때 rawValue를 사용하지 않도록 수정
- description을 사용할 수 있도록 CustomStringConvertible 프로토콜 채택
- 상기 수정에 따른 기존 rawValue를 description으로 대체

llimental committed:
[refactor] banker work() 메서드 변경

- 기존: dequeue된 현재 customer의 waitingNumber와 workType을 파라미터로 받아 기능 구현
- 변경: dequeue된 현재 customer를 파라미터로 받아 기능 구현
- 그 외 currentCustomerWorkType이 많이 쓰이지 않아 제거
- switch문에서 조건을 currentCustomer의 workType을 사용하도록 변경

calledBlu committed:
[refactor] Bank 초기화 시 Banker가 주입될 수 있도록 리팩토링

- 생성자 주입을 통하여 Bank가 생성될 때 각각의 banker 프로퍼티에 workType에 맞는 Banker가 생성되도록 리팩토링

llimental committed:
[feat] workTime 측정을 위한 timeInterval 사용

- 비동기 작업 시작 전 startDate로 현재 시간 저장
- 비동기 작업 종료 후 timeIntervalSince 를 통해 startDate와의 차이 측정
- 측정한 elapsedTime을 매개변수로 넘겨 전체 코드 작동 시간을 출력

calledBlu committed:
[docs] Update STEP 3 PR link
